File: ruby/lib/coding_interview/domain/stock.rb
require "bigdecimal"

module CodingInterview
  module Domain
    # Stock は保有銘柄（銘柄と保有額）を表す。
    Stock = Struct.new(:symbol, :amount_jpy)

    # PortfolioItem は最適ポートフォリオの1銘柄エントリー（銘柄と構成比率）を表す。
    PortfolioItem = Struct.new(:symbol, :rate)

    # Portfolio は最適ポートフォリオ（銘柄ごとの構成比率）を表す。
    class Portfolio
      attr_reader :items

      def initialize(items)
        raise ArgumentError, "portfolio must have at least one item" if items.empty?
        total = items.map(&:rate).inject(BigDecimal("0")) { |acc, r| acc + r }
        raise ArgumentError, "portfolio rates must sum to 1, got #{total}" unless total == BigDecimal("1")
        symbols = items.map(&:symbol)
        raise ArgumentError, "portfolio must not have duplicate symbols" if symbols.uniq.size != symbols.size
        @items = items.freeze
      end
    end

  end
end

File: python/src/coding_interview/domain/stock.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from coding_interview.domain.stock_symbol import StockSymbol


# Stock は保有銘柄（銘柄と保有額）を表す。
@dataclass(frozen=True)
class Stock:
    symbol: StockSymbol
    amount_jpy: Decimal


# PortfolioItem はポートフォリオの銘柄ごとの構成比率を表す。
@dataclass(frozen=True)
class PortfolioItem:
    symbol: StockSymbol
    rate: Decimal


# Portfolio は最適ポートフォリオ（銘柄ごとの構成比率）を表す。
@dataclass(frozen=True)
class Portfolio:
    items: tuple[PortfolioItem, ...]

    def __post_init__(self) -> None:
        if not self.items:
            raise ValueError("portfolio must have at least one item")
        rate_sum = sum(item.rate for item in self.items)
        if rate_sum != Decimal(1):
            raise ValueError(f"portfolio rates must sum to 1, got {rate_sum}")
        symbols = [item.symbol for item in self.items]
        if len(symbols) != len(set(symbols)):
            raise ValueError("portfolio must not have duplicate symbols")
